### DISCUSSION TITLE
Cannot use text boxes under wayland

### DISCUSSION THREAD

**Discussion by GiantNuker** [ppy/osu#14976] [Q&A] [answered]
Cannot type in text boxes (login, map search) under Wayland (swaywm), while game keys and menu keyboard navigation work. Text boxes work fine under Xorg (i3wm)

System: Arch Linux (everything is updated), SwayWM Wayland compositor
Osu Version: 2021.1005.0-lazer

Dosen't appear to be very useful but heres the [runtime.log](https://github.com/ppy/osu/files/7289831/runtime.log)

**Comment by pizzadude** [comment]
same issue on fedora 35 kde wayland

**Comment by bdach** [comment, marked as answer]
Please wait for the next release and test that. We've merged an SDL bump that may resolve some Wayland issues.

That said that compositor is not something any of the active contributors are actively using to my knowledge and therefore there's really no promises with regard to support.

**Reply by GiantNuker** [reply]
Dosn't seem to have fixed it.

Would it be possible to handle text box input the same way as keybinds? Those work.

**Reply by bdach** [reply]
unfortunate.

obviously text boxes being broken is bad but it requires someone to go see what the cause is on wayland. i am not familiar with that compositor and have no idea why text would be broken there, and with all due respect that is not exactly the most pressing issue given the existing platforms (windows) take precedence and there is an available workaround.

**Reply by GiantNuker** [reply]
makes sense, i might look into it myself later, but at least its not that big of an issue

**Reply by GiantNuker** [reply]
it seems the event is at least recieved, checking to see where it stops
![20211028_12h19m09s_grim](https://user-images.githubusercontent.com/43734540/139321462-0605a5b3-b353-405d-9316-eb4a81072024.png)

-- Edit 1
It appears an SDL_TEXTINPUT event is never submitted, while SDL_KEYUP and SDL_KEYDOWN both are.

-- Edit 2
SDL_TEXTINPUT would seem to give unicode input, while KEYUP/DOWN will only give ascii

**Reply by GiantNuker** [reply]
WOOO
![20211028_13h51m42s_grim](https://user-images.githubusercontent.com/43734540/139333862-0fcd686d-dabb-41f8-b42c-88051cb0a617.png)

SDL_StartTextInput() was never called!
[https://wiki.libsdl.org/Tutorials-TextInput](https://wiki.libsdl.org/Tutorials-TextInput)
[https://wiki.libsdl.org/SDL_StartTextInput](https://wiki.libsdl.org/SDL_StartTextInput)

as car as i can tell this is the root cause of the problem @bdach 
I'll make a PR when i find a good spot to put the event (instead of spamming it randomly)

**Reply by bdach** [reply]
You should check out https://github.com/ppy/osu-framework/pull/4709 first. It likely will do the same things that you need it to to fix this.

Probably need to prioritise reviewing that if it does fix...

**Reply by GiantNuker** [reply]
I'll clone it and check, from reading the pr discussion, it seems like it probably will

**Reply by GiantNuker** [reply]
@bdach that PR does fix it (i only pulled framework, as i couldnt get it working with osu, tested the text input box in the framework tests)
![20211028_15h32m03s_grim](https://user-images.githubusercontent.com/43734540/139345250-ad25b92f-d9aa-447e-a004-ca34e0089f97.png)

**Reply by bdach** [reply]
thanks for testing. I'll look into prioritising it for review and merge.

**Reply by GiantNuker** [reply]
no problem, thanks

**Comment by pizzadude** [comment]
@bdach @GiantNuker I found a workaround:

env -u WAYLAND_DISPLAY ./osu.AppImage (to force it to run in xwayland)

This fixes the issue for now.

**Reply by GiantNuker** [reply]
i'll test it and see if xwayland adds any noticeable lag
its still better than starting an x session in another tty just to search beatmaps though

**Reply by GiantNuker** [reply]
dosen't even launch for me (with DRI_PRIME=1 (osu-lazer-dgpu) or not (osu-lazer))
![20211026_23h04m07s_grim](https://user-images.githubusercontent.com/43734540/139008964-147501d2-e4a0-49c2-927a-fb7ac3140383.png)
Its probably because im missing some random package, or because im using lazer from the AUR and not the appimage directly

**Reply by pizzadude** [reply]
try the appimage. also i'm on intel graphics (thinkpad). and i use kde not sway.

**Reply by GiantNuker** [reply]
nope, same result (amd graphics here)

**Reply by pizzadude** [reply]
idk then

**Reply by ekrctb** [reply]
`env -u WAYLAND_DISPLAY` is weird and suboptimal. Try
```
SDL_VIDEODRIVER=x11 osu-lazer
```
instead. If that works, you may have `SDL_VIDEODRIVER` set. See if `echo $SDL_VIDEODRIVER` prints anything.

**Reply by GiantNuker** [reply]
yep that one works

**Comment by pizzadude** [comment]
The latest osu release seems to fix the issue.

**Reply by GiantNuker** [reply]
probably related to my weird configuration then

just to be safe, are you sure its not launching with X? (i checked with xeyes, that only works when launching in windowed mode though)
if it is launching under wayland i wonder if kde's compositor is doing something special

**Reply by pizzadude** [reply]
I don't think xeyes works with games that capture the mouse. I could be wrong.

All I know is I don't have to use the env var anymore.

**Reply by GiantNuker** [reply]
Xeyes worked fine for me as long as i set SDL_VIDEODRIVER=x11 and launch the game in windowed mode

**Reply by bdach** [reply]
well... yeah. from the little i know i'm relatively sure `xeyes` works for X applications and on wayland it only works for windows that go through xwayland.

on proper native wayland applications/windows `xeyes` is expected to _not_ work.

**Reply by GiantNuker** [reply]
A more reliable way to check ive found is
`xwininfo -tree -root | grep osu!`
Under X it should look something like this:
![20211028_12h10m24s_grim](https://user-images.githubusercontent.com/43734540/139320299-5870f8df-4bc8-4f37-9129-3e27cf87300e.png)
Under XWayland it should be empty like this:
![20211028_12h11m47s_grim](https://user-images.githubusercontent.com/43734540/139320473-78dfe1ec-d613-43a7-9765-4745a9b50727.png)


~~Thats the point~~
~~If its running under xwayland then xeyes should look at the window, if its running under native wayland it wont follow the mouse in the wayland window~~

**Reply by pizzadude** [reply]
Ah. Looks like it is running under XWayland.

**Reply by GiantNuker** [reply]
You probably have the environment variable removed in your bashrc or something, the root cause of the problem is above if you care

**Reply by pizzadude** [reply]
> You probably have the environment variable removed in your bashrc or something, the root cause of the problem is above if you care

I didn't modify my bashrc.
